sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -618,6 +619,52 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 /**
+ * Propagate foldable expressions:
+ * Replace all attributes with aliases of the original foldable expressions except Union queries.
+ * Aliases and ordinal expressions are the main target to be transformed after propagation. Other
+ * optimizations will take advantage of the propagated foldable expressions.
+ *
+ * {{{
+ *   SELECT 1.0 x, 'abc' y, Now() z ORDER BY x, y, 3
+ *   ==>  SELECT 1.0 x, 'abc' y, Now() z ORDER BY 1.0, 'abc', Now()
+ * }}}
+ */
+object FoldablePropagation extends Rule[LogicalPlan] {
+  private def hasCommandOrUnion(plan: LogicalPlan): Boolean = plan.find {
+    case _: Command => true
+    case _: Union => true
+    case _ => false
+  }.isDefined
+
+  private def replaceInSubplan(plan: LogicalPlan): LogicalPlan = plan.transformUp {
+    case subplan: LogicalPlan if !hasCommandOrUnion(subplan) => this(subplan)
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    val foldableExprSet = ExpressionSet(plan.flatMap {
+      case Project(projectList, _) => projectList.collect {
+        case a: Alias if a.resolved && a.child.foldable => a
+      }
+      case _ => Nil
+    })
+
+    if (foldableExprSet.isEmpty) {
+      plan
+    } else if (hasCommandOrUnion(plan)) {
+      replaceInSubplan(plan)
```

Comment:
Do you mean changing all the current optimizer logic or only `replaceInSubplan`?